great_expectations/rule_based_profiler/config/base.py

```diff
@@ -0,0 +1,214 @@
+
+
+@dataclass(frozen=True)
+class DomainBuilderConfig(DictDot):
+    class_name: str
+    module_name: Optional[str] = None
+    batch_request: Optional[Dict[str, Any]] = None
+
+
+class DomainBuilderConfigSchema(NotNullSchema):
+    class Meta:
+        unknown = INCLUDE
+
+    __config__ = DomainBuilderConfig
+
+    class_name = fields.String(required=True)
+    module_name = fields.String(
+        required=False,
+        all_none=True,
+        missing="great_expectations.rule_based_profiler.domain_builder",
+    )
+    batch_request = fields.Dict(keys=fields.String(), required=False, allow_none=True)
+
+
+@dataclass(frozen=True)
+class ParameterBuilderConfig(DictDot):
+    name: str
+    class_name: str
+    module_name: Optional[str] = None
+    batch_request: Optional[Dict[str, Any]] = None
+
+
+class ParameterBuilderConfigSchema(NotNullSchema):
+    class Meta:
+        unknown = INCLUDE
+
+    __config__ = ParameterBuilderConfig
+
+    name = fields.String(required=True)
+    class_name = fields.String(required=True)
+    module_name = fields.String(
+        required=False,
+        all_none=True,
+        missing="great_expectations.rule_based_profiler.parameter_builder",
+    )
+    batch_request = fields.Dict(keys=fields.String(), required=False, allow_none=True)
+
+
+@dataclass(frozen=True)
+class ExpectationConfigurationBuilderConfig(DictDot):
+    expectation_type: str
+    class_name: str
+    module_name: Optional[str] = None
+    mostly: Optional[float] = None
+    meta: Optional[Dict] = None
+
+
+class ExpectationConfigurationBuilderConfigSchema(NotNullSchema):
+    class Meta:
+        unknown = INCLUDE
+
+    __config__ = ExpectationConfigurationBuilderConfig
+
+    class_name = fields.String(required=True)
+    module_name = fields.String(
+        required=False,
+        all_none=True,
+        missing="great_expectations.rule_based_profiler.expectation_configuration_builder",
+    )
+    expectation_type = fields.String(required=True)
+    mostly = fields.Float(required=False, allow_none=True)
+    meta = fields.Dict(required=False, allow_none=True)
+
+
+@dataclass(frozen=True)
+class RuleConfig(DictDot):
+    name: str
+    domain_builder: DomainBuilderConfig
+    parameter_builders: List[ParameterBuilderConfig]
+    expectation_configuration_builders: List[ExpectationConfigurationBuilderConfig]
+
+
+class RuleConfigSchema(NotNullSchema):
+    class Meta:
+        unknown = INCLUDE
+
+    __config__ = RuleConfig
+
+    name = fields.String(required=True)
+    domain_builder = fields.Nested(DomainBuilderConfigSchema, required=True)
+    parameter_builders = fields.List(
+        cls_or_instance=fields.Nested(ParameterBuilderConfigSchema, required=True),
+        required=True,
+    )
+    expectation_configuration_builders = fields.List(
+        cls_or_instance=fields.Nested(
+            ExpectationConfigurationBuilderConfigSchema, required=True
+        ),
+        required=True,
+    )
+
+
+@dataclass
```

Comment:
Need to remove the frozen nature of the class if we want to invoke the superclass constructor in `__post_init__`.

The alternative is to do the work of the superclass constructor with `setattr` but it feels wrong to duplicate parent class logic. Are we okay with making this mutable? How can we balance immutability and an appropriate child-parent relationship?

Reply:
@cdkini I am not sure if we should make any of these configurations mutable.  A big part of the reconciliation logic is to detect `batch_data`, delete the value, then serialize/deserialize (to validate the configuration against the Marshmallow Schema), and then add the `batch_data` back.  All these operations are on the configuration objects -- they extend `DictDot` to enable this easily.  Hence, for this reason, they should be mutable.  Thanks.  /cc @donaldheppner

Reply:
Open to get Don's input here. Don't love getting away from immutable objects but will reconsider based on use-case.